int32 horizontal
int32 vertical